File: typescript/src/server/authority-server.ts
import { Controller, GET } from '../spec'
import type {
  ClientSideServerAnyAuthorityRecord,
  ClientSideServerAnyTrustedServerRecord,
  ClientSideServerExportedKeyPair,
  ClientSideServerSignedAuthorityRecord,
} from '../client-side-server/identity'
import {
  createSignedClientSideServerAuthorityRecord,
  createSignedClientSideServerAuthorityRecordV2,
  isClientSideServerTrustedServerRecordV2,
} from '../client-side-server/identity'

export interface AuthorityHostSummary {
  serverName: string
  keyId?: string
  fingerprint: string
  source: string
  trustedAt: number
}

export interface AuthorityHostListResponse {
  authorityName?: string
  total: number
  hosts: AuthorityHostSummary[]
}

export interface AuthorityHostExportResponse {
  authorityName?: string
  total: number
  records: ClientSideServerAnyAuthorityRecord[]
}

export interface PLATAuthorityServerOptions {
  authorityName?: string
  authorityKeyPair: ClientSideServerExportedKeyPair
  knownHosts: Record<string, ClientSideServerAnyTrustedServerRecord | ClientSideServerAnyTrustedServerRecord[]>
  allowServerNames?: string[]
  allow?: (serverName: string, record: ClientSideServerAnyTrustedServerRecord) => boolean
}

export function createAuthorityServerController(
  options: PLATAuthorityServerOptions,
): new () => any {
  const authorityName = options.authorityName

  @Controller('authority')
  class AuthorityServerApi {
    @GET()
    async resolveAuthorityHost(
      { serverName }: { serverName: string },
    ): Promise<ClientSideServerAnyAuthorityRecord | ClientSideServerAnyAuthorityRecord[] | null> {
      const records = getAuthorityHostRecords(options, serverName)
      if (records.length === 0) return null
      if (records.length === 1) {
        return signAuthorityHostRecord(options, serverName, records[0]!, authorityName)
      }
      return Promise.all(records.map((record) => signAuthorityHostRecord(options, serverName, record, authorityName)))
    }

    @GET()
    async listAuthorityHosts(
      input: { q?: string; limit?: number; offset?: number; serverNames?: string[] } = {},
    ): Promise<AuthorityHostListResponse> {
      const records = selectAuthorityHostRecords(options, input)
      return {
        authorityName,
        total: records.length,
        hosts: records.map(([serverName, record]) => ({
          serverName,
          keyId: isClientSideServerTrustedServerRecordV2(record) ? record.signingKeyId : record.keyId,
          fingerprint: isClientSideServerTrustedServerRecordV2(record) ? record.signingFingerprint : record.fingerprint,
          source: record.source,
          trustedAt: record.trustedAt,
        })),
      }
    }

    @GET()
    async exportAuthorityHosts(
      input: { q?: string; limit?: number; offset?: number; serverNames?: string[] } = {},
    ): Promise<AuthorityHostExportResponse> {
      const records = selectAuthorityHostRecords(options, input)
      return {
        authorityName,
        total: records.length,
        records: await Promise.all(
            records.map(async ([serverName, record]) => signAuthorityHostRecord(options, serverName, record, authorityName)),
        ),
      }
    }
  }

  return AuthorityServerApi
}

function getAuthorityHostRecords(
  options: PLATAuthorityServerOptions,
  serverName: string,
): ClientSideServerAnyTrustedServerRecord[] {
  const entry = options.knownHosts[serverName]
  if (!entry) return []
  if (options.allowServerNames && !options.allowServerNames.includes(serverName)) return []
  const records = Array.isArray(entry) ? entry : [entry]
  return records.filter((record) => !options.allow || options.allow(serverName, record))
}

function selectAuthorityHostRecords(
  options: PLATAuthorityServerOptions,
  input: { q?: string; limit?: number; offset?: number; serverNames?: string[] },
): Array<[string, ClientSideServerAnyTrustedServerRecord]> {
  const pairs: Array<[string, ClientSideServerAnyTrustedServerRecord]> = []
  for (const [serverName, entry] of Object.entries(options.knownHosts)) {
    const records = getAuthorityHostRecords(options, serverName)
    for (const record of records) {
      pairs.push([serverName, record])
    }
  }

  const filtered = pairs
    .filter(([serverName]) => !input.serverNames?.length || input.serverNames.includes(serverName))
    .filter(([serverName]) => !input.q || serverName.toLowerCase().includes(input.q.toLowerCase()))
    .sort(([left], [right]) => left.localeCompare(right))

  const offset = Math.max(0, input.offset ?? 0)
  const limit = Math.max(1, input.limit ?? (filtered.length || 1))
  return filtered.slice(offset, offset + limit)
}

async function signAuthorityHostRecord(
  options: PLATAuthorityServerOptions,
  serverName: string,
  record: ClientSideServerAnyTrustedServerRecord,
  authorityName?: string,
): Promise<ClientSideServerSignedAuthorityRecord | ClientSideServerAnyAuthorityRecord> {
  return isClientSideServerTrustedServerRecordV2(record)
    ? createSignedClientSideServerAuthorityRecordV2(options.authorityKeyPair, {
        serverName,
        signingPublicKeyJwk: record.signingPublicKeyJwk,
        encryptionPublicKeyJwk: record.encryptionPublicKeyJwk,
        signingKeyId: record.signingKeyId,
        encryptionKeyId: record.encryptionKeyId,
        authorityName,
      })
    : createSignedClientSideServerAuthorityRecord(options.authorityKeyPair, {
        serverName,
        publicKeyJwk: record.publicKeyJwk,
        keyId: record.keyId,
        authorityName,
      })
}

File: python/plat/authority_server.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Callable

from .css_identity import CSSAuthorityKeyPair, CSSKnownHostRecord, create_signed_authority_record
from .decorators import Controller, GET


@dataclass
class AuthorityServerOptions:
    authority_key_pair: CSSAuthorityKeyPair
    known_hosts: dict[str, CSSKnownHostRecord]
    authority_name: str | None = None
    allow_server_names: list[str] | None = None
    allow: Callable[[str, CSSKnownHostRecord], bool] | None = None


def create_authority_server_controller(options: AuthorityServerOptions):
    @Controller("authority")
    class AuthorityServerApi:
        @GET()
        def resolveAuthorityHost(self, serverName: str):
            record = _get_host_record(options, serverName)
            if record is None:
                return None
            return _serialize_record(
                create_signed_authority_record(
                    options.authority_key_pair,
                    server_name=serverName,
                    public_key_jwk=record.public_key_jwk,
                    key_id=record.key_id,
                    authority_name=options.authority_name,
                )
            )

        @GET()
        def listAuthorityHosts(
            self,
            q: str | None = None,
            limit: int | None = None,
            offset: int | None = None,
            serverNames: list[str] | None = None,
        ):
            records = _select_host_records(options, q=q, limit=limit, offset=offset, server_names=serverNames)
            return {
                "authorityName": options.authority_name,
                "total": len(records),
                "hosts": [
                    {
                        "serverName": server_name,
                        "keyId": record.key_id,
                        "fingerprint": record.fingerprint,
                        "source": record.source,
                        "trustedAt": record.trusted_at,
                    }
                    for server_name, record in records
                ],
            }

        @GET()
        def exportAuthorityHosts(
            self,
            q: str | None = None,
            limit: int | None = None,
            offset: int | None = None,
            serverNames: list[str] | None = None,
        ):
            records = _select_host_records(options, q=q, limit=limit, offset=offset, server_names=serverNames)
            return {
                "authorityName": options.authority_name,
                "total": len(records),
                "records": [
                    _serialize_record(
                        create_signed_authority_record(
                            options.authority_key_pair,
                            server_name=server_name,
                            public_key_jwk=record.public_key_jwk,
                            key_id=record.key_id,
                            authority_name=options.authority_name,
                        )
                    )
                    for server_name, record in records
                ],
            }

    return AuthorityServerApi


def _get_host_record(options: AuthorityServerOptions, server_name: str) -> CSSKnownHostRecord | None:
    record = options.known_hosts.get(server_name)
    if record is None:
        return None
    if options.allow_server_names and server_name not in options.allow_server_names:
        return None
    if options.allow and not options.allow(server_name, record):
        return None
    return record


def _select_host_records(
    options: AuthorityServerOptions,
    *,
    q: str | None = None,
    limit: int | None = None,
    offset: int | None = None,
    server_names: list[str] | None = None,
) -> list[tuple[str, CSSKnownHostRecord]]:
    filtered = [
        (server_name, record)
        for server_name, record in sorted(options.known_hosts.items())
        if _get_host_record(options, server_name) is record
        and (not server_names or server_name in server_names)
        and (not q or q.lower() in server_name.lower())
    ]
    start = max(0, offset or 0)
    size = max(1, limit or len(filtered) or 1)
    return filtered[start : start + size]


def _serialize_record(record) -> dict:
    return {
        "protocol": record.protocol,
        "serverName": record.server_name,
        "publicKeyJwk": record.public_key_jwk,
        "keyId": record.key_id,
        "authorityName": record.authority_name,
        "issuedAt": record.issued_at,
        "signature": record.signature,
    }
